lib/livebook_web/live/session_live.ex

```diff
@@ -788,4 +788,38 @@ defmodule LivebookWeb.SessionLive do
       changed?: info.evaluation_digest != nil and info.digest != info.evaluation_digest
     }
   end
+
+  # Updates current data_view in response to an operation.
+  # In most cases we simply recompute data_view, but for the
+  # most common ones we only update the relevant parts.
+  defp update_data_view(data_view, data, operation) do
+    case operation do
+      {:report_cell_revision, _pid, _cell_id, _revision} ->
+        data_view
+
+      {:apply_cell_delta, _pid, cell_id, _delta, _revision} ->
+        update_cell_view(data_view, data, cell_id)
+
+      _ ->
+        data_to_view(data)
+    end
+  end
+
+  defp update_cell_view(data_view, data, cell_id) do
+    {:ok, cell, section} = Notebook.fetch_cell_and_section(data.notebook, cell_id)
+
+    update_in(data_view, [:section_views, Access.all()], fn section_view ->
+      if section_view.id == section.id do
+        update_in(section_view, [:cell_views, Access.all()], fn cell_view ->
+          if cell_view.id == cell.id do
+            cell_to_view(cell, data)
+          else
+            cell_view
+          end
+        end)
+      else
+        section_view
+      end
+    end)
+  end
```

Comment:
If you want to be fancy, you could create an Access-like function called `access_by_id(id)` so you could do this:

    put_in data_view[:section_views, access_by_id(section.id), :cell_views, access_by_id(cell.id)], cell_to_view(cell, data)

It can be useful if you could use it elsewhere in the app. But this current version also works!

Reply:
I like that! We have similar nested updates for the `Notebook` struct, so I took this as an opportunity to clean it up as well. Let me know if the implementation makes sense =)